### PR TITLE
chore(watcher): test burn log handling via mock

### PR DESCRIPTION
**Pull request by TheAncientGoat**
This PR increases watcher test coverage, by allowing us to mock the fetching of burn event logs

### CODE REVIEW

**Review by jazg — APPROVED** [review]
Looks great, big fan of the restructure!

**Comment by jazg** [comment]
TODO:
- Add any additional test cases for catching edge/failure cases
- Send error logs to Sentry for further debugging

File: watcher/watcher_test.go [hunk 1]
```diff
@@ -2,26 +2,58 @@ package watcher_test
...
 
+	"github.com/ethereum/go-ethereum/core/types"
 	_ "github.com/mattn/go-sqlite3"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	v0 "github.com/renproject/lightnode/compat/v0"
 	. "github.com/renproject/lightnode/watcher"
 
-	"github.com/alicebob/miniredis"
+	"github.com/alicebob/miniredis/v2"
```
Comment (tok-kkk) [inline watcher/watcher_test.go:16]: Nice. That's something I'm looking for.

File: watcher/watcher.go [hunk 2]
```diff
@@ -25,35 +25,90 @@ import (
...
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+	resultChan := make(chan BurnLogResult)
+
+	go func() {
+		for iter.Next() {
+			err := iter.Error()
```
Comment (tok-kkk) [inline watcher/watcher.go:66]: I think the error should be checked outside the `iter.Next()` 
```golang
for iter.Next(){
      // ...
}
if iter.Error() != nil {
     // ...
}

```
Also I notice there's `iter.Close()` function we can use. Might not be the reason causing the issue, but could release some resources as a small optimization.

File: watcher/watcher.go [hunk 3]
```diff
@@ -25,35 +25,90 @@ import (
...
+			if err != nil {
+				resultChan <- BurnLogResult{Error: err}
+				break
+			}
+			resultChan <- BurnLogResult{Result: *iter.Event}
+			select {
+			case <-ctx.Done():
+				break
```
Comment (tok-kkk) [inline watcher/watcher.go:74]: This `break` only breaks the `select` statement, not the `for` loop.

**Review by tok-kkk — CHANGES_REQUESTED** [review]
Looks good to me.

File: watcher/watcher.go (hunk 3 above)
Comment (TheAncientGoat) [reply watcher/watcher.go:74]: Good catch

File: watcher/watcher.go (hunk 2 above)
Comment (TheAncientGoat) [reply watcher/watcher.go:66]: Doesn't the iter close when it's done anyhow?

Comment (tok-kkk) [reply watcher/watcher.go:66]: Doesn't look like it. The binding creates a `event.Subscription` when initializing the iterator. I can't see it get `Unsubscribe` anywhere expect the `iter.Close()` function. I think we can do something like
```golang
iter, err := bindings.FilterLogBurn()
if err != nil {
    // ...
}
defer iter.Close()
```
Event the subscription gets unsubscribed elsewhere, it should be fine as the actual implementation uses `sync.Once` to ignore extra calls.

Comment (TheAncientGoat) [reply watcher/watcher.go]: Ok, addressed this
